omc3/knob_extractor.py

```diff
@@ -0,0 +1,459 @@
+        outfile.write(f"!! --- extracted knobs for time {time}\n\n")
+        for category, knobs in category_knobs.items():
+            if not knobs:
+                continue
+            outfile.write(f"!! --- {category:10} --------------------\n")
+            for knob, knob_entry in knobs.items():
+                outfile.write(f"{knob_entry.get_madx()}\n")
+            outfile.write("\n")
+        outfile.write("\n")
+
+
+# Knobs Dict -------------------------------------------------------------------
+
+def _get_knobs_def_file(user_defined: Optional[Union[Path, str]] = None) -> Path:
+    """ Check which knobs-definition file is appropriate to take. """
+    if user_defined is not None:
+        LOGGER.info(f"Using user defined knobs.txt: '{user_defined}")
+        return Path(user_defined)
+
+    if KNOBS_TXT_MDLDIR.is_file():
+        LOGGER.info(f"Using model folder's knobs.txt: '{KNOBS_TXT_MDLDIR}")
+        return KNOBS_TXT_MDLDIR
+
+    if KNOBS_TXT_AFS.is_file():
+        # if all fails, fall back to lhc acc-models
+        LOGGER.info(f"Using fallback knobs.txt: '{KNOBS_TXT_AFS}'")
+        return KNOBS_TXT_AFS
+
+    raise FileNotFoundError("None of the knobs-definition files are available.")
+
+
+def _load_knobs_dict(file_path: Union[Path, str]) -> KnobsDict:
+    """ Load the knobs-definition file and convert into KnobsDict.
+    Each line in this file should consist of four comma separated entries:
+    madx-name, lsa-name, scaling factor, knob-test value.
+    Alternatively, a TFS-file is also allowed, but needs to have the suffix ``.tfs``.
+
+    Args:
+        file_path (Path): Path to the knobs definition file.
+
+    Returns:
+        Dictionary with LSA names (but with colon instead of /) as
+        keys and KnobEntries (without values) as value.
+    """
+    if Path(file_path).suffix == ".tfs":
+        # just in case someone wants to give tfs files (hidden feature)
+        df = tfs.read_tfs(file_path)
+    else:
+        # parse csv file (the official way)
+        dtypes = {"madx": str, "lsa": str, "scaling": float, "test": float}
+        converters = {'madx': str.strip, 'lsa': str.strip}  # strip whitespaces
+        df = pd.read_csv(file_path, comment="#", names=dtypes.keys(), dtype=dtypes, converters=converters)
+    return _dataframe_to_knobsdict(df)
+
+
+def _dataframe_to_knobsdict(df: pd.DataFrame) -> KnobsDict:
+    """ Converts a DataFrame into the required Dictionary structure.
+
+    Args:
+        df (pd.DataFrame): DataFrame containing at least the columns
+                           'lsa', 'madx', 'scaling' (upper or lowercase)
+
+    Returns:
+        Dictionary with LSA names (but with colon instead of /) as
+        keys and KnobEntries (without values) as value.
+
+    """
+    df.columns = df.columns.astype(str).str.lower()
+    df = df[['lsa', 'madx', 'scaling']].set_index("lsa", drop=False)
+    return {
+        lsa2name(r[0]): KnobEntry(**r[1].to_dict()) for r in df.iterrows()
+    }
+
```

Comment:
Again, this is nice to have, no question. But there is no intention to use such a functionality

Reply:
What do you mean? It's convenient for anything that stays in the Python process and doesn't want to go through command line level.